crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1267 @@
+            Self::move_version_from_dirty_to_cache(
+                &self.dirty.markers,
+                &self.cached.marker_cache,
+                (epoch, object_key.0),
+                object_key.1,
+                marker_value,
+            );
+        }
+
+        for (object_id, object) in written.iter() {
+            Self::move_version_from_dirty_to_cache(
+                &self.dirty.objects,
+                &self.cached.object_cache,
+                *object_id,
+                object.version(),
+                &ObjectEntry::Object(object.clone()),
+            );
+        }
+
+        for ObjectKey(object_id, version) in deleted.iter() {
+            Self::move_version_from_dirty_to_cache(
+                &self.dirty.objects,
+                &self.cached.object_cache,
+                *object_id,
+                *version,
+                &ObjectEntry::Deleted,
+            );
+        }
+
+        for ObjectKey(object_id, version) in wrapped.iter() {
+            Self::move_version_from_dirty_to_cache(
+                &self.dirty.objects,
+                &self.cached.object_cache,
+                *object_id,
+                *version,
+                &ObjectEntry::Wrapped,
+            );
+        }
+
+        let tx_digest = *transaction.digest();
+        let effects_digest = effects.digest();
+
+        self.dirty
+            .transaction_effects
+            .remove(&effects_digest)
+            .expect("effects must exist");
+
+        match self.dirty.transaction_events.entry(events.digest()) {
+            DashMapEntry::Occupied(mut occupied) => {
+                let txns = &mut occupied.get_mut().0;
+                assert!(txns.remove(&tx_digest), "transaction must exist");
+                if txns.is_empty() {
+                    occupied.remove();
+                }
+            }
+            DashMapEntry::Vacant(_) => {
+                panic!("events must exist");
+            }
+        }
+
+        self.dirty
+            .executed_effects_digests
+            .remove(&tx_digest)
+            .expect("executed effects must exist");
+
+        Ok(())
+    }
+
+    // Move the oldest/least entry from the dirty queue to the cache queue.
+    // This is called after the entry is committed to the db.
+    fn move_version_from_dirty_to_cache<K, V>(
+        dirty: &DashMap<K, CachedVersionMap<V>>,
+        cache: &MokaCache<K, Arc<Mutex<CachedVersionMap<V>>>>,
+        key: K,
+        version: SequenceNumber,
+        value: &V,
+    ) where
+        K: Eq + std::hash::Hash + Clone + Send + Sync + Copy + 'static,
+        V: Send + Sync + Clone + Eq + std::fmt::Debug + 'static,
+    {
+        static MAX_VERSIONS: usize = 3;
+
+        // IMPORTANT: lock both the dirty set entry and the cache entry before modifying either.
+        // this ensures that readers cannot see a value temporarily disappear.
+        let cache_entry = cache.entry(key).or_default();
```

Comment:
Can't believe I missed this. I thought I was doing dirty then cached everywhere.

Reply:
I tried writing a mut version of `with_locked_cache_entries`, but it doesn't end up saving any code. Its quite different from the non-mut version because: a) we need to assert that the dirty entry is occupid b) create the cached entry if it doesn't exist and c) remove the dirty entry from the map if it is empty when we are done`